### PR TITLE
Raytrace entities within collection

### PR DESCRIPTION
**Pull request by NeumimTo**
Adds method `World#raytrace` that takes a collection of entities instead of Predicate<Entity>
..another way to filter out entities we dont't want to check for ray intersection

### CODE REVIEW

File: Spigot-API-Patches/0231-Raytrace-entities-within-collection.patch [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
++     * Performs a ray trace that checks for entity collisions within predefined collection of entities
++     * <p>
++     *
++     * @param direction the ray direction
++     * @param maxDistance the maximum distance
++     * @param raySize entity bounding boxes will be uniformly expanded (or
++     *    shrinked) by this value before doing collision checks
++     * @param startPos the start position
++     * @param entities Collection of entities to check against
++     * @return the closest ray trace hit result, or <code>null</code> if there
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Raytrace-entities-within-collection.patch]:
```suggestion
+     * @return the closest ray trace hit result, or {@code null} if there
```

File: Spigot-API-Patches/0231-Raytrace-entities-within-collection.patch [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
++     *   is no hit
++     */
++    @Nullable
++    public RayTraceResult rayTraceEntities(@NotNull Vector direction, double maxDistance, double raySize, @NotNull Vector startPos, @NotNull Collection<Entity> entities);
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Raytrace-entities-within-collection.patch]:
```suggestion
+    public RayTraceResult rayTraceEntities(@NotNull Vector direction, double maxDistance, double raySize, @NotNull Vector startPos, @NotNull Collection<? extends Entity> entities);
```

File: Spigot-API-Patches/0231-Raytrace-entities-within-collection.patch [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+
+diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
+index e827e1a6f5c0f8410ed32dda8f17aa769b469999..b7fed2399a3dceb120da3a90da875d81f394d66a 100644
+--- a/src/main/java/org/bukkit/World.java
++++ b/src/main/java/org/bukkit/World.java
+@@ -1659,6 +1659,22 @@ public interface World extends PluginMessageRecipient, Metadatable {
+     @Nullable
+     public RayTraceResult rayTraceEntities(@NotNull Location start, @NotNull Vector direction, double maxDistance, double raySize, @Nullable Predicate<Entity> filter);
+
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Raytrace-entities-within-collection.patch:14]: Paper comment

File: Spigot-Server-Patches/0594-Raytrace-entities-within-collection.patch [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+index 299f57ca2a65887a0d7e7c584fc1bd5c783b0db2..f6fe354948f7aa6fdfa22c6bcb20818c7d768961 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+@@ -1304,6 +1304,11 @@ public class CraftWorld implements World {
+         BoundingBox aabb = BoundingBox.of(startPos, startPos).expandDirectional(dir).expand(raySize);
+         Collection<Entity> entities = this.getNearbyEntities(aabb, filter);
+ 
++        return rayTraceEntities(direction, maxDistance, raySize, startPos, entities);
```
Comment (Proximyst) [inline Spigot-Server-Patches/0594-Raytrace-entities-within-collection.patch:16]: Paper comment

File: Spigot-Server-Patches/0594-Raytrace-entities-within-collection.patch [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
++    @Override
++    public RayTraceResult rayTraceEntities(Vector direction, double maxDistance, double raySize, Vector startPos, Collection<Entity> entities) {
+         Entity nearestHitEntity = null;
+         RayTraceResult nearestHitResult = null;
+         double nearestDistanceSq = Double.MAX_VALUE;
+@@ -1326,6 +1331,7 @@ public class CraftWorld implements World {
+         return (nearestHitEntity == null) ? null : new RayTraceResult(nearestHitResult.getHitPosition(), nearestHitEntity, nearestHitResult.getHitBlockFace());
+     }
+ 
++
```
Comment (Proximyst) [inline Spigot-Server-Patches/0594-Raytrace-entities-within-collection.patch]: Remove this

File: Spigot-API-Patches/0231-Raytrace-entities-within-collection.patch [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+index e827e1a6f5c0f8410ed32dda8f17aa769b469999..b7fed2399a3dceb120da3a90da875d81f394d66a 100644
+--- a/src/main/java/org/bukkit/World.java
++++ b/src/main/java/org/bukkit/World.java
+@@ -1659,6 +1659,22 @@ public interface World extends PluginMessageRecipient, Metadatable {
+     @Nullable
+     public RayTraceResult rayTraceEntities(@NotNull Location start, @NotNull Vector direction, double maxDistance, double raySize, @Nullable Predicate<Entity> filter);
+ 
++    /**
++     * Performs a ray trace that checks for entity collisions within predefined collection of entities
++     * <p>
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Raytrace-entities-within-collection.patch]: Why? There is no new paragraph after this.

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0231-Raytrace-entities-within-collection.patch (hunk 3 above)
Comment (NeumimTo) [reply Spigot-API-Patches/0231-Raytrace-entities-within-collection.patch:14]: what does that mean?

Comment (Proximyst) [reply Spigot-API-Patches/0231-Raytrace-entities-within-collection.patch:14]: Check the contributing file

File: Spigot-API-Patches/0231-Raytrace-entities-within-collection.patch [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
+index e827e1a6f5c0f8410ed32dda8f17aa769b469999..d1d259236d9def64cec68c45a4de007fc1f335c9 100644
+--- a/src/main/java/org/bukkit/World.java
++++ b/src/main/java/org/bukkit/World.java
+@@ -1659,6 +1659,24 @@ public interface World extends PluginMessageRecipient, Metadatable {
+     @Nullable
+     public RayTraceResult rayTraceEntities(@NotNull Location start, @NotNull Vector direction, double maxDistance, double raySize, @Nullable Predicate<Entity> filter);
+ 
++    // Paper start
++
```
Comment (Proximyst) [inline Spigot-API-Patches/0231-Raytrace-entities-within-collection.patch]: Drop this line.

File: Spigot-Server-Patches/0594-Raytrace-entities-within-collection.patch [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+         Collection<Entity> entities = this.getNearbyEntities(aabb, filter);
+ 
++        // Paper start
++        return rayTraceEntities(direction, maxDistance, raySize, startPos, entities);
++        // Paper end
++    }
++
++    // Paper start
++    @Override
++    public RayTraceResult rayTraceEntities(Vector direction, double maxDistance, double raySize, Vector startPos, Collection<? extends Entity> entities) {
```
Comment (Proximyst) [inline Spigot-Server-Patches/0594-Raytrace-entities-within-collection.patch]:
```suggestion
+        // Paper start - add entities collection
+        return rayTraceEntities(direction, maxDistance, raySize, startPos, entities);
+    }
+
+    @Override
+    public RayTraceResult rayTraceEntities(Vector direction, double maxDistance, double raySize, Vector startPos, Collection<? extends Entity> entities) {
+        // Paper end
```

File: Spigot-Server-Patches/0594-Raytrace-entities-within-collection.patch [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+         RayTraceResult nearestHitResult = null;
+         double nearestDistanceSq = Double.MAX_VALUE;
+@@ -1325,6 +1333,7 @@ public class CraftWorld implements World {
+ 
+         return (nearestHitEntity == null) ? null : new RayTraceResult(nearestHitResult.getHitPosition(), nearestHitEntity, nearestHitResult.getHitBlockFace());
+     }
++    // Paper end
+ 
+     @Override
+     public RayTraceResult rayTraceBlocks(Location start, Vector direction, double maxDistance) {
```
Comment (Proximyst) [inline Spigot-Server-Patches/0594-Raytrace-entities-within-collection.patch]:
```suggestion
```

**Review by Proximyst — DISMISSED** [review]
Nitpicks; LGTM

File: Spigot-Server-Patches/0594-Raytrace-entities-within-collection.patch [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+index 299f57ca2a65887a0d7e7c584fc1bd5c783b0db2..9ee28222d8dbb3564d00e4ed18c4865ce5a62874 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+@@ -1304,6 +1304,13 @@ public class CraftWorld implements World {
+         BoundingBox aabb = BoundingBox.of(startPos, startPos).expandDirectional(dir).expand(raySize);
+         Collection<Entity> entities = this.getNearbyEntities(aabb, filter);
+ 
++        // Paper start - add entities collection
++        return rayTraceEntities(direction, maxDistance, raySize, startPos, entities);
```
Comment (electronicboy) [inline Spigot-Server-Patches/0594-Raytrace-entities-within-collection.patch:16]: This is basically missing all of the sanity verifications which are done in the method above

**Review by electronicboy — CHANGES_REQUESTED** [review]

**Pull request closed without merging**